models/ecoli/analysis/variant/ribosome_crowding.py

```diff
@@ -0,0 +1,168 @@
+				early_cell_mask = early_cell_mask[exclude_timeout_cell_mask]
+				avg_overcrowded_monomer_indexes_early = np.where(
+					sum((avg_actual_prob_translation_per_transcript <
+						avg_target_prob_translation_per_transcript)[
+							early_cell_mask]) > 0)[0]
+
+				late_cell_mask = all_generations[variant] >= MIN_LATE_CELL_INDEX
+				if len(late_cell_mask) == 1:
+					late_cell_mask = late_cell_mask[0]
+				late_cell_mask = late_cell_mask[exclude_timeout_cell_mask]
+				avg_overcrowded_monomer_indexes_late = np.where(
+					sum((avg_actual_prob_translation_per_transcript <
+						avg_target_prob_translation_per_transcript)[
+							late_cell_mask]) > 0)[0]
+
+			# Record that these genes were overcrowded this variant
+			for monomer_index in avg_overcrowded_monomer_indexes:
+				if monomer_index not in overcrowded_monomer_index_dicts[0]:
+					overcrowded_monomer_index_dicts[0][monomer_index] = []
+				overcrowded_monomer_index_dicts[0][monomer_index].append(
+					variant)
+
+			if exclude_early_gens == 1:
+				for monomer_index in avg_overcrowded_monomer_indexes_early:
+					if monomer_index not in overcrowded_monomer_index_dicts[1]:
+						overcrowded_monomer_index_dicts[1][monomer_index] = []
+					overcrowded_monomer_index_dicts[1][monomer_index].append(
+						variant)
+
+				for monomer_index in avg_overcrowded_monomer_indexes_late:
+					if monomer_index not in overcrowded_monomer_index_dicts[2]:
+						overcrowded_monomer_index_dicts[2][monomer_index] = []
+					overcrowded_monomer_index_dicts[2][monomer_index].append(
+						variant)
+
+		# Determine the gene ids corresponding to these proteins
+		with open(simDataFile, 'rb') as f:
+			sim_data = pickle.load(f)
+		mRNA_sim_data = sim_data.process.transcription.cistron_data.struct_array
+		monomer_sim_data = sim_data.process.translation.monomer_data.struct_array
+		monomer_to_mRNA_id_dict = dict(zip(monomer_sim_data['id'],
+										   monomer_sim_data[
+											   'cistron_id']))
+		mRNA_to_gene_id_dict = dict(zip(mRNA_sim_data['id'],
+										mRNA_sim_data['gene_id']))
+
+		overcrowded_monomer_indexes = overcrowded_monomer_index_dicts[0].keys()
+		overcrowded_gene_indexes_to_ids_dict = {monomer_index:
+			mRNA_to_gene_id_dict.get(monomer_to_mRNA_id_dict.get(
+				monomer_ids[monomer_index])) for monomer_index in
+			overcrowded_monomer_indexes}
+
+		# Output
+		print("---Writing Output---")
+		output_descr = ["_all_gens"]
+		if exclude_early_gens == 1:
+			output_descr += ["_early_gens", "_late_gens"]
+
+		for j in range(len(overcrowded_monomer_index_dicts)):
+			with open(os.path.join(plotOutDir, plotOutFileName +
+				output_descr[j] + ".txt"), 'w') as f:
+				for monomer_id in overcrowded_monomer_index_dicts[j]:
+					overcrowded_variant_string = ", ".join(map(str,
+						overcrowded_monomer_index_dicts[j][monomer_id]))
+					f.write("" + overcrowded_gene_indexes_to_ids_dict[
+						monomer_id] + ": " + overcrowded_variant_string + "\n")
```

Comment:
Using [f-strings](https://www.geeksforgeeks.org/formatted-string-literals-f-strings-python/) might be a more cleaner/flexible option for formatting these types of strings.